### ISSUE TITLE
Mass descriptions: Comparative mass descriptions for BASICS - size & density

### ISSUE THREAD

**Issue by terracoda** [phetsims/gravity-force-lab-basics#96]
Comparative description of the size of the sphere is provided in the REGULAR sim, see https://github.com/phetsims/gravity-force-lab/issues/146.

This description is only provided in dynamic `aria-valuetext` on focus of one of the mass sliders.

I think a similar comparative description in BASICS will be useful, but the comparative scale needs adjusting due to the huge scales (billions of kilograms) used in this sim.

There are only 10 values for mass in this sim.

Assigning to myself to create a scale for implementation.

**Comment by terracoda** [comment]
@zepumph, as we discussed last Tuesday/Wednesday, I have re-worked the scale for qualitative comparative size, and also created a matching scale for qualitative comparative density.

- Comparative size descriptions are used in relevant places when Constant Size checkbox is not checked. 
- Comparative density descriptions are used in the same relevant places when Constant Size check box is checked.

We also discussed that hopefully we could use the same scale for both REGULAR and BASICS. I think we can, but we might have to tweak the ranges a bit for REGULAR as there are many more values for mass in that sim. Regardless of REGULAR, I think we can start BASICS off will the following ranges in this 9-point scale.

Link to the [Qualitative Scales](https://docs.google.com/document/d/1HdDG9ds2MdbCb21l9qk3cI8yBQxK6-wBWNXC4Tloji8/edit#heading=h.yc2oelnnexw8) section of the design document. There are a few tables that explore the descriptions side by side with visual representations. 

This is the "Tweaked" version of the scale. I'll update it if anything changes.
<img width="886" alt="new-scales-size-density" src="https://user-images.githubusercontent.com/1902116/53735039-00b65580-3e86-11e9-9bd8-f57a14d5f680.png">

**Note** that in the design doc, I did explore whether "much smaller"/"much less dense" should be greater than or less than "half the size of". The table above is the"Tweaked version" where they are bigger, but would love your thoughts, on this.

**Comment by terracoda** [comment]
@zepumph, I realize I have not highlighted the relevant places in the PDOM and alerts that are affected by these descriptions

The first order of business is dynamic `aria-valuetext` for the mass sliders on focus as you succeeded in doing here https://github.com/phetsims/gravity-force-lab/issues/146#issuecomment-464966560 for REGULAR.

**Comment by terracoda** [comment]
With an effort to keep issues short and focused. I'll make other issues for the other descriptions and alerts related to toggling the Constant Size checkbox.

**Comment by terracoda** [comment]
@zepumph, I am adding an example from the design doc for the values read out on focus of a mass stepper.
**Currently implemented we have no described comparison of mass:**
- {{2}} billion kilograms, mass 1 stepper. Force arrows {{tiny}}. 

Adding a dynamic comparative description in the `aria-valuetext` on focus only for the mass stepper. The comparison (size or density) would sound like this:
**With Constant Size NOT checked:**
- {{2}} billion kilograms, {{twice the size of}} {{mass 2}}, mass 1 stepper. Force arrows {{tiny}}.
**With Constant Size checked:**
- {{2}} billion kilograms, {{twice as dense as}} {{mass 2}}, mass 1 stepper. Force arrows {{tiny}}.

**Comment by zepumph** [comment]
I was not able to implement https://github.com/phetsims/gravity-force-lab-basics/issues/96#issuecomment-469265259 just yet. I will look into https://github.com/phetsims/sun/issues/482 to support this, as well as to improve the slight workaround/hack in GFL/MassControl that is very related to this.

**Comment by terracoda** [comment]
@zepumph, it is ok that we might not have comparative mass and comparative density descriptions (on focus of the stepper) until you address https://github.com/phetsims/sun/issues/482

We hope to have interviews again in early April, so it would be great to get the values for mass/density values working (see #109) and then we can deal with the comparative mass/density descriptions on focus later. 

Note that as soon as the toggling of mass/density descriptions is working (#106) the comparative mass/density descriptions will both be in the PDOM, so it's not like they are completely missing.

**Comment by zepumph** [comment]
Once this issue is finished, we will be ready to review https://github.com/phetsims/scenery/issues/951.

**Comment by zepumph** [comment]
I was able to tackle. https://github.com/phetsims/sun/issues/482 (aka phetsims/scenery#951) today. After doing that this issue was very straight forward (as expected since that was the reason for the common code work!). @terracoda will you please review the "on focus" valuetext changes for both REGULAR and BASICS, and let me know if there is anything else here. 

Note that the constant size checkbox has not been wired up to anything yet. But I hope to get to that next!

**Comment by zepumph** [comment]
Sorry, It is actually this issue where I will add the density comparisons now.

**Comment by zepumph** [comment]
Here is a patch that almost adds the relative density descriptions, but it isn't totally working, so I don't want to commit just yet. I hope to work on this again before the week is over though.

<details>

```
Index: js/gravity-force-lab/view/describers/MassDescriber.js
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- js/gravity-force-lab/view/describers/MassDescriber.js	(revision cf4d4318090fbc1a90ae0bb3c332818d0abd3d25)
+++ js/gravity-force-lab/view/describers/MassDescriber.js	(date 1553646054305)
@@ -42,6 +42,16 @@
   const twiceTheSizeOfString = GravityForceLabA11yStrings.twiceTheSizeOf.value;
   const muchMuchLargerThanString = GravityForceLabA11yStrings.muchMuchLargerThan.value;
 
+  const notDenseComparedToString = GravityForceLabA11yStrings.notDenseComparedTo.value;
+  const halfAsDenseAsString = GravityForceLabA11yStrings.halfAsDenseAs.value;
+  const muchLessDenseThanString = GravityForceLabA11yStrings.muchLessDenseThan.value;
+  const lessDenseButComparableToString = GravityForceLabA11yStrings.lessDenseButComparableTo.value;
+  const asDenseAsString = GravityForceLabA11yStrings.asDenseAs.value;
+  const denseButComparableToString = GravityForceLabA11yStrings.denseButComparableTo.value;
+  const muchDenseThanString = GravityForceLabA11yStrings.muchDenseThan.value;
+  const twiceAsDenseAsString = GravityForceLabA11yStrings.twiceAsDenseAs.value;
+  const extremelyDenseComparedToString = GravityForceLabA11yStrings.extremelyDenseComparedTo.value;
+
   const massChangeClausePatternString = GravityForceLabA11yStrings.massChangeClausePattern.value;
   const massChangesAndMovesClausePatternString = GravityForceLabA11yStrings.massChangesAndMovesClausePattern.value;
   const massChangesMovesOtherClausePatternString = GravityForceLabA11yStrings.massChangesMovesOtherClausePattern.value;
@@ -72,6 +82,17 @@
     twiceTheSizeOfString,
     muchMuchLargerThanString
   ];
+  const RELATIVE_DENSITY_STRINGS = [
+    notDenseComparedToString,
+    halfAsDenseAsString,
+    muchLessDenseThanString,
+    lessDenseButComparableToString,
+    asDenseAsString,
+    denseButComparableToString,
+    muchDenseThanString,
+    twiceAsDenseAsString,
+    extremelyDenseComparedToString
+  ];
 
   const { OBJECT_ONE } = ISLCObjectEnum;
 
@@ -99,6 +120,9 @@
       this.convertMassValue = options.convertMassValue;
       this.formatMassValue = options.formatMassValue;
 
+      // @private
+      this.constantRadiusProperty = model.constantRadiusProperty;
+
       model.object1.valueProperty.link( ( newMass, oldMass ) => {
         this.mass1Growing = ( newMass - oldMass ) > 0;
       } );
@@ -113,7 +137,7 @@
      * @returns {string}
      */
     getMassValuesSummaryText() {
-      const relativeSize = this.getRelativeSize( OBJECT_ONE );
+      const relativeSize = this.getRelativeSizeOrDensity( OBJECT_ONE );
 
       return StringUtils.fillIn( massValuesAndComparisonSummaryPatternString, {
         mass1Label: this.object1Label,
@@ -129,15 +153,15 @@
      * @returns {string}
      */
     getM1RelativeSize() {
-      return this.getMassRelativeSizeToOther( OBJECT_ONE );
+      return this.getMassRelativeSizeOrDensityToOther( OBJECT_ONE );
     }
 
     /**
      * @param {ISLCObjectEnum} objectEnum
      * @returns {string}
      */
-    getMassRelativeSizeToOther( objectEnum ) {
-      const relativeSize = this.getRelativeSize( objectEnum );
+    getMassRelativeSizeOrDensityToOther( objectEnum ) {
+      const relativeSize = this.getRelativeSizeOrDensity( objectEnum );
       const firstObjectLabel = this.getObjectLabelFromEnum( objectEnum );
       const secondObjectLabel = this.getOtherObjectLabelFromEnum( objectEnum );
       return StringUtils.fillIn( objectsRelativeSizePatternString, {
@@ -257,6 +281,15 @@
       return RELATIVE_SIZE_STRINGS[ index ];
     }
 
+    /**
+     * @param {number} index - should be an index
+     * @returns {*}
+     */
+    getRelativeDensityFromIndex( index ) {
+      assert && assert( Util.isInteger( index ) );
+      return RELATIVE_DENSITY_STRINGS[ index ];
+    }
+
 
     /**
      * Returns the mapped integer corresponding to the appropriate qualitative size comparison between masses.
@@ -298,12 +331,18 @@
       assert && assert( false, `unrecognized ratio: ${ratio}` );
     }
 
-    getRelativeSize( thisObjectEnum ) {
+    /**
+     * @param {ISLCObjectEnum} thisObjectEnum
+     * @returns {string}
+     */
+    getRelativeSizeOrDensity( thisObjectEnum ) {
       const thisObject = this.getObjectFromEnum( thisObjectEnum );
       const otherObject = this.getOtherObjectFromEnum( thisObjectEnum );
       const ratio = thisObject.valueProperty.value / otherObject.valueProperty.value;
       const index = this.getRelativeSizeIndex( ratio );
-      return this.getRelativeSizeFromIndex( index );
+
+      // use size or density depending on if constant checkbox is checked.
+      return this.constantRadiusProperty.get() ? this.getRelativeDensityFromIndex( index ) : this.getRelativeSizeFromIndex( index );
     }
 
     /**
@@ -344,7 +383,7 @@
      */
     getVerboseMassAriaValueText( objectEnum ) {
       const massAndUnit = this.getMassAndUnit( objectEnum );
-      const relativeSize = this.getRelativeSize( objectEnum );
+      const relativeSize = this.getRelativeSizeOrDensity( objectEnum );
       const otherObjectLabel = this.getOtherObjectLabelFromEnum( objectEnum );
       return StringUtils.fillIn( massValueRelativeSizePatternString, {
         massAndUnit: massAndUnit,
Index: js/gravity-force-lab/GravityForceLabA11yStrings.js
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- js/gravity-force-lab/GravityForceLabA11yStrings.js	(revision cf4d4318090fbc1a90ae0bb3c332818d0abd3d25)
+++ js/gravity-force-lab/GravityForceLabA11yStrings.js	(date 1553646018900)
@@ -123,6 +123,35 @@
       value: 'gets bigger'
     },
 
+    // relative densities
+    notDenseComparedTo: {
+      value: 'not dense compared to'
+    },
+    halfAsDenseAs: {
+      value: 'half as dense as'
+    },
+    muchLessDenseThan: {
+      value: 'much less dense than'
+    },
+    lessDenseButComparableTo: {
+      value: 'less dense but comparable to'
+    },
+    asDenseAs: {
+      value: 'as dense as'
+    },
+    denseButComparableTo: {
+      value: 'denser but comparable to'
+    },
+    muchDenseThan: {
+      value: 'much denser than'
+    },
+    twiceAsDenseAs: {
+      value: 'twice as dense as'
+    },
+    extremelyDenseComparedTo: {
+      value: 'extremely dense compared to'
+    },
+
     // mass sphere density
     notAtAllDense: {
       value: 'not at all dense'

```

</details>

**Comment by zepumph** [comment]
Alright I think this is all working now. The mass control aria-valuetext attributes update correctly, and other pieces of the PDOM seem to be updating as expected too. @terracoda will you please review now.

**Comment by terracoda** [comment]
Looks and sounds good @zepumph! Closing this issue and already closed #106.
